### PR TITLE
feat(ivy): add canonical example of a pipe.

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: packages/core/src/render3/pipe.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * Invokes a pure pipe with 4 arguments.
+ *
+ * This instruction acts as a guard to {@link PipeTransform#transform} invoking
+ * the pipe only when an input to the pipe changes.
+ *
+ * @param index Pipe index where the pipe was stored on creation.
+ * @param v1 1st argument to {@link PipeTransform#transform}.
+ */
+export function pipeBind1(index: number, v1: any): any {
+  throw new Error('TODO: implement!');
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/pipe.ts]: Those functions will be very similar to the existing `b0...bV` (logic to compare arguments, memorize those etc.). I wanted to refactor those methods so split them up into arguments checking / memorization before - maybe this is a good occasion?

**Comment by pkozlowski-opensource** [comment]
One things I'm missing here is an example showing lifecycle hook(s) on pipes - if I'm not mistaken we can at least have `OnDestroy`.

Other than this LGTM as a general idea. 
We will see more details when the runtime part gets implemented.

**Review by pkozlowski-opensource — DISMISSED** [review]
Not sure if this is WIP and you are also planning to add:
- runtime impl
- lifecycle hooks 
in this or a separate PR.

As a very minimum this PR needs re-format - lint is failing

**Comment by mhevery** [comment]
I was not planning on doing any implementation since this PR was to unblock @chuckjaz on complier implementation. Perhaps you or someone from Amadeus can take on the implementation?

I will fix:
- [ ] fix formating
- [ ] Add Lifecycle hooks.

File: packages/core/src/render3/pipe.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/render3/pipe.ts]: Yes we should do this. Also chuck makes a good point that most of the code gen is in form of `"{{exp}}"' which translates to `b1('', exp, '')`. So we should probably have it generate `b1(exp)` and than have separate instruction to add wrap like so `w('', b1(exp),'')` and only generate the wrap if needed.

**Comment by mhevery** [comment]
@pkozlowski-opensource please have another look.

File: packages/core/src/metadata/directives.ts [hunk 2]
```diff
@@ -783,13 +783,36 @@ export interface PipeDecorator {
  * @stable
  */
 export interface Pipe {
+  /**
+   * Name of the pipe.
+   *
+   * The pipe name is used in template bindings. For example if a pipe is named
+   * `myPipe` than it would be used in the template binding expression like
```
Comment (kara) [inline packages/core/src/metadata/directives.ts]: Typo: than -> then

File: packages/core/src/metadata/directives.ts [hunk 3]
```diff
@@ -783,13 +783,36 @@ export interface PipeDecorator {
...
+   * so:  `{{ exp | myPipe }}`.
+   */
   name: string;
+
+  /**
+   * If Pipe is pure (its output depends only on its input.)
+   *
+   * Normally pipe's `translate` method is invoked on each changed detection
```
Comment (kara) [inline packages/core/src/metadata/directives.ts]: Typo: changed -> change

File: packages/core/src/metadata/directives.ts [hunk 4]
```diff
@@ -783,13 +783,36 @@ export interface PipeDecorator {
...
+   * cycle. If the cost of invoking the `translated` method is non-trivial and
+   * if the output of the pipe depends only on its input, than declaring a pipe
```
Comment (kara) [inline packages/core/src/metadata/directives.ts]: Typo: than -> then

File: packages/core/src/metadata/directives.ts [hunk 5]
```diff
@@ -783,13 +783,36 @@ export interface PipeDecorator {
...
+   * as pure would short circuit the invocation of the `translate` method and
+   * invoke the method only when the inputs to the pipe change.
+   */
   pure?: boolean;
 }
 
 /**
  * Pipe decorator and metadata.
  *
+ * Use the `@Pipe` annotation to declare the a given class is a pipe. A pipe
```
Comment (kara) [inline packages/core/src/metadata/directives.ts]: Typo: declare the a -> declare that a

File: packages/core/src/render3/definition.ts [hunk 6]
```diff
@@ -149,11 +150,33 @@ function invertObject(obj: any): any {
...
     DirectiveDef<T>;
+
+/**
+ * Create a pipe definition object.
+ *
+ * # Example
+ * ```
+ * class MyPipe implements PipeTransform {
+ *   // Generated by Angular Template Compiler
+ *   static ngPipDef = definePipe({
```
Comment (kara) [inline packages/core/src/render3/definition.ts]: Typo: ngPipDef -> ngPipeDef

File: packages/core/src/render3/interfaces/definition.ts [hunk 7]
```diff
@@ -106,6 +109,31 @@ export interface ComponentDef<T> extends DirectiveDef<T> {
...
+  /**
+   * factory function used to create a new directive instance.
+   *
+   * NOTE: this property is short (1 char) because it is used in
+   * component templates which is sensitive to size.
+   */
+  n: () => PipeTransform;
+
+  /**
+   * Wether or not the pipe is pure.
```
Comment (kara) [inline packages/core/src/render3/interfaces/definition.ts]: Typo: Wether -> Whether

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 8]
```diff
@@ -240,6 +241,69 @@ describe('compiler specification', () => {
...
+      name: 'myPurePipe',
+      pure: true,
+    })
+    class MyPurePipe implements PipeTransform {
+      transform(value: any, ...args: any[]) { throw new Error('Method not implemented.'); }
+
+      // NORMATIVE
+      static ngPipeDef = r3.definePipe({
+        type: MyPurePipe,
+        factory: function MyPipe_Factory() { return new MyPipe(); },
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical_spec.ts]: I think this should be

```ts
factory: function MyPurePipe_Factory() { return new MyPurePipe(); }
```

**Review by kara — COMMENTED** [review]
Mostly typos

File: packages/core/src/metadata/directives.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/metadata/directives.ts]: 👍

File: packages/core/src/metadata/directives.ts (hunk 3 above)
Comment (mhevery) [reply packages/core/src/metadata/directives.ts]: 👍

File: packages/core/src/metadata/directives.ts (hunk 4 above)
Comment (mhevery) [reply packages/core/src/metadata/directives.ts]: 👍

File: packages/core/src/metadata/directives.ts (hunk 5 above)
Comment (mhevery) [reply packages/core/src/metadata/directives.ts]: 👍

File: packages/core/src/render3/definition.ts (hunk 6 above)
Comment (mhevery) [reply packages/core/src/render3/definition.ts]: 👍

File: packages/core/src/render3/interfaces/definition.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/src/render3/interfaces/definition.ts]: 👍

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical_spec.ts]: 👍

File: packages/core/src/render3/pipe.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * Invokes a pure pipe with 4 arguments.
+ *
+ * This instruction acts as a guard to {@link PipeTransform#transform} invoking
+ * the pipe only when an input to the pipe changes.
+ *
+ * @param index Pipe index where the pipe was stored on creation.
+ * @param v1 1st argument to {@link PipeTransform#transform}.
+ * @param v2 1st argument to {@link PipeTransform#transform}.
```
Comment (kara) [inline packages/core/src/render3/pipe.ts]: I think all of these in this file say "1st argument". This one should be 2nd, no?

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 10]
```diff
@@ -240,6 +241,69 @@ describe('compiler specification', () => {
...
+      name: 'myPurePipe',
+      pure: true,
+    })
+    class MyPurePipe implements PipeTransform {
+      transform(value: any, ...args: any[]) { throw new Error('Method not implemented.'); }
+
+      // NORMATIVE
+      static ngPipeDef = r3.definePipe({
+        type: MyPurePipe,
+        factory: function MyPurePipe_Factory() { return new MyPipe(); },
```
Comment (kara) [inline packages/core/test/render3/compiler_canonical_spec.ts]: Should be `new MyPurePipe();`, right?

File: packages/core/src/render3/pipe.ts (hunk 9 above)
Comment (mhevery) [reply packages/core/src/render3/pipe.ts]: fixed

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 10 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical_spec.ts]: fixed

**Review by kara — APPROVED** [review]
LGTM

**Pull request closed without merging**